src/main/scala/org/broadinstitute/dsde/rawls/expressions/ExpressionParsing.scala

```diff
@@ -0,0 +1,224 @@
+
+  // add pipe to an entity referenced by the current entity
+  private def entityNameAttributePipeFunc(entityRefName: String)(pipe: PipeType): PipeResult = {
+    PipeResult(
+      // an entity name is on the outgoing edge label
+      pipe.out(entityRefName),
+      // TODO: is there a way to automate returning this code-string from the code itself? (used for debugging purposes)
+      s".out($entityRefName)"
+    )
+  }
+
+  // TODO: We only support attributes here, not reference types (edges)
+  private def lastAttributePipeFunc(attributeName: String)(pipe: PipeType): FinalResult = {
+    val lastVertices = pipe.toList
+
+    if (lastVertices.size() == 0) {
+      throw new RuntimeException(s"Could not dereference $attributeName because pipe returned no entities")
+    }
+
+    // Look up the attributes on the list of vertices returned from the pipe. This will insert nulls into the list if
+    // the attribute doesn't exist.
+    FinalResult(lastVertices.map((v: Vertex) => {
+      v.getProperty(attributeName).asInstanceOf[String]
+    }), s".getProperty($attributeName)")
+  }
+}
+
+class ExpressionEvaluator(graph:Graph, parser:ExpressionParser)  {
+  def evaluate(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, expression:String):Try[Seq[Any]] = {
+    parser.parse(expression)(workspaceNamespace, workspaceName, rootType, rootName) match {
+      case Success(pipe) => runPipe(workspaceNamespace, workspaceName, rootType, rootName, pipe)
+      case Failure(regret) => Failure(regret)
+    }
+  }
+
+  def runPipe(workspaceNamespace:String, workspaceName:String, rootType:String, rootName:String, steps:PipelineQuery):Try[Seq[Any]] = {
+    val pipeline = new PipeType(graph)
+
+    // TODO: more idiomatic way to do this?  the GremlinPipeline has side effects, and adds pipes to each .out() and similar calls anyway, so maybe not...
+    var pipelineQueryText = ""
+    var result:PipeResult = null
+
+    // apply the root function, this will give us the root vertex we will traverse from
+    result = steps.rootFunc(pipeline, workspaceNamespace, workspaceName, rootType, rootName)
+    pipelineQueryText += result.pipeAction
+
+    // then if we have any pipes between root and last, apply each all but last pipes from this root
+    steps.pipes.foreach(pipes => {
+      for (step <- pipes) {
+        result = step(pipeline)
+
+        pipelineQueryText += result.pipeAction
+      }
+    })
+
+    // apply the last action
+    steps.lastAction match {
+      // if there
+      case Some(last) => {
+        try {
+          val finalResult = last(pipeline)
+          pipelineQueryText += finalResult.pipeAction
+          //println("DEBUG: running query: " + pipelineQueryText)
+          Success(finalResult.result)
+        } catch {
+          case rxc : RuntimeException => Failure(rxc)
+        }
+      }
+      // evaluate root-only expression (i.e. an entity. TODO: think about representation?)
+      case None => {
```

Comment:
See comment [here](https://github.com/broadinstitute/rawls/pull/37/files#r32991663); we'll want to be able to return entities here too.